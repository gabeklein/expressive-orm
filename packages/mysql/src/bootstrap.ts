import { Field, Type, Util } from '@expressive/sql';

export function bootstrap(entities: Iterable<Type.EntityType>){
  const commands = [] as string[];
  
  for(const entity of entities)
    entity.ready();

  for(const entity of entities)
    commands.push(create(entity));

  for(const entity of entities){
    const statement = constraints(entity);

    if(statement)
      commands.push(statement);
  }
  
  return commands.join(";");
}

export function drop(table: Type.EntityType){
  return `DROP TABLE IF EXISTS ${table.name}`
}

export function create(table: Type.EntityType){
  const { table: name } = table;
  const statements = [] as string[];

  table.fields.forEach(field => {
    const sql = column(field);

    if(sql)
      statements.push(sql);
  });

  return `CREATE TABLE IF NOT EXISTS ${name} (${statements.join(",")})`
}

export function constraints(table: Type.EntityType){
  const statement = [] as string[];

  table.fields.forEach(field => {
    const { constraint } = field;

    if(constraint)
      statement.push(constraint);
  })

  if(statement.length)
    return `ALTER TABLE ${table.name} ${statement.join(", ")}`
}

function column(from: Field){
  if(from.datatype === undefined)
    return;

  let statement = `${Util.qualify(from.column)} ${from.datatype}`;

  if(!from.nullable)
    statement += " NOT NULL";

  if(from.default !== undefined)
    statement += ` DEFAULT ${Util.escapeString(from.default)}`;

  if(from.datatype == "INT" && from.increment)
    statement += " AUTO_INCREMENT";

  if(from.primary)
    statement += " PRIMARY KEY";

  return statement;
}