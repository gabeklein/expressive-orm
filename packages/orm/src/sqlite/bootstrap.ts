import Entity from '..';
import Field from '../Field';
import { escapeString, qualify } from '../utility';

export function bootstrap(entities: Iterable<Entity.Type>){
  const commands = [] as string[];
  
  for(const entity of entities)
    entity.ensure();

  for(const entity of entities)
    commands.push(table(entity));

  for(const entity of entities){
    const statement = constraint(entity);

    if(statement)
      commands.push(statement);
  }
  
  return commands.join(";");
}

export function drop(tables: Entity.Type[]){
  const commands = [];

  for(const table of tables)
    commands.push(`DROP TABLE IF EXISTS ${table.name}`);

  return commands;
}

export function table(table: Entity.Type){
  const { table: name } = table;
  const statements = [] as string[];

  table.fields.forEach(field => {
    const sql = column(field);

    if(sql)
      statements.push(sql);
  });

  return `CREATE TABLE IF NOT EXISTS ${name} (${statements.join(",")})`;
}

export function constraint(table: Entity.Type){
  const statement = [] as string[];

  table.fields.forEach(field => {
    const { constraint } = field;

    if(constraint)
      statement.push(constraint);
  })

  if(statement.length)
    return `ALTER TABLE ${table.name} ${statement.join(", ")}`;
}

export function column(from: Field){
  let { datatype } = from;
  let statement = qualify(from.column);

  if(datatype === undefined)
    return;

  if(from.primary){
    if(!datatype.includes("INT"))
      throw new Error("Primary key may only be INTEGER.")

    datatype = "INTEGER";
    statement += " INTEGER PRIMARY KEY";

    if(from.increment !== false)
      from.increment = true;
  }
  else {
    statement += ` ${datatype}`;

    if(!from.nullable)
      statement += " NOT NULL";
  
    if(from.default !== undefined)
      statement += ` DEFAULT ${escapeString(from.default)}`;
  }
  
  if(from.increment){
    if(datatype == "INTEGER")
      statement += " AUTOINCREMENT";
    else
      throw new Error("Can only auto-increment INTEGER values.")
  }

  return statement;
}