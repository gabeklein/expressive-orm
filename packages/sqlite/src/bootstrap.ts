import { Type, Field, Util } from '@expressive/sql';

export function bootstrap(entities: Iterable<Type.EntityType>){
  const commands = [] as string[];
  
  for(const entity of entities)
    entity.ready();

  for(const entity of entities)
    commands.push(table(entity));

  for(const entity of entities){
    const statement = constraint(entity);

    if(statement)
      commands.push(statement);
  }
  
  return commands.join(";");
}

export function drop(tables: Type.EntityType[]){
  const commands = [];

  for(const table of tables)
    commands.push(`DROP TABLE IF EXISTS ${table.name}`);

  return commands;
}

export function table(table: Type.EntityType){
  const { table: name } = table;
  const statements = [] as string[];

  table.fields.forEach(field => {
    const sql = column(field);

    if(sql)
      statements.push(sql);
  });

  return `CREATE TABLE IF NOT EXISTS ${name} (${statements.join(",")})`;
}

export function constraint(table: Type.EntityType){
  const statement = [] as string[];

  table.fields.forEach(field => {
    const { constraint } = field;

    if(constraint)
      statement.push(constraint);
  })

  if(statement.length)
    return `ALTER TABLE ${table.name} ${statement.join(", ")}`;
}

export function column(from: Field){
  let { datatype } = from;
  let statement = Util.qualify(from.column);

  if(datatype === undefined)
    return;

  if(from.primary){
    if(!datatype.includes("INT"))
      throw new Error("Primary key may only be INTEGER.")

    datatype = "INTEGER";
    statement += " INTEGER PRIMARY KEY";

    if(from.increment !== false)
      from.increment = true;
  }
  else {
    statement += ` ${datatype}`;

    if(!from.nullable)
      statement += " NOT NULL";
  
    if(from.default !== undefined)
      statement += ` DEFAULT ${Util.escapeString(from.default)}`;
  }
  
  if(from.increment){
    if(datatype == "INTEGER")
      statement += " AUTOINCREMENT";
    else
      throw new Error("Can only auto-increment INTEGER values.")
  }

  return statement;
}