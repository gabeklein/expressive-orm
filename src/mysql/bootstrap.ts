import Entity from '..';
import Field from '../Field';
import { escapeString, qualify } from '../utility';
import MySQLConnection from './Connection';

function bootstrap(connection: MySQLConnection, dryRun: true): string; 
function bootstrap(connection: MySQLConnection, dryRun?: false): Promise<void>;
function bootstrap(connection: MySQLConnection, dryRun?: boolean): string | Promise<void>;
function bootstrap(connection: MySQLConnection, dryRun?: boolean){
  const tables = Array.from(connection.managed.values());
  const commands = [] as string[];

  if(connection.options.nuke)
    commands.push(...drop(tables));

  commands.push(...create(tables));
  commands.push(...constraints(tables))
  
  const sql = commands.join(";");

  return dryRun ? sql : connection.query(sql);
}

export default bootstrap;

function drop(tables: Entity.Type[]){
  const commands = [];

  for(const table of tables)
    commands.push(`DROP TABLE IF EXISTS ${table.name}`);

  return commands;
}

function create(tables: Entity.Type[]){
  const commands = [];

  for(const table of tables){
    const { table: name } = table;
    const statements = [] as string[];

    table.fields.forEach(field => {
      const sql = column(field);

      if(sql)
        statements.push(sql);
    });

    commands.push(
      `CREATE TABLE IF NOT EXISTS ${name} (${statements.join(",")})`
    )
  }

  return commands;
}

function constraints(tables: Entity.Type[]){
  const commands = [] as string[];

  for(const table of tables){
    const statement = [] as string[];

    table.fields.forEach(field => {
      const { constraint } = field;

      if(constraint)
        statement.push(constraint);
    })

    if(statement.length)
      commands.push(`ALTER TABLE ${table.name} ${statement.join(", ")}`);
  }

  return commands;
}

function column(from: Field){
  if(from.datatype === undefined)
    return;

  let statement = `${qualify(from.column)} ${from.datatype}`;

  if(!from.nullable)
    statement += " NOT NULL";

  if(from.default !== undefined)
    statement += ` DEFAULT ${escapeString(from.default)}`;

  if(from.datatype == "INT" && from.increment)
    statement += " AUTO_INCREMENT";

  if(from.primary)
    statement += " PRIMARY KEY";

  return statement;
}