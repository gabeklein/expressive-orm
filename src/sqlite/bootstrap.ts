import Entity, { Connection } from '..';
import Field from '../Field';
import { escapeString, qualify } from '../utility';

function bootstrap(connection: Connection, dryRun: true): string; 
function bootstrap(connection: Connection, dryRun?: false): Promise<void>;
function bootstrap(connection: Connection, dryRun?: boolean): string | Promise<void>;
function bootstrap(connection: Connection, dryRun?: boolean){
  const tables = Array.from(connection.managed.values());
  const commands = [] as string[];

  if(false)
    commands.push(...drop(tables));

  commands.push(...create(tables));
  commands.push(...constraints(tables))
  
  const sql = commands.join(";");

  return dryRun ? sql : connection.query(sql);
}

export default bootstrap;

function drop(tables: Entity.Type[]){
  const commands = [];

  for(const table of tables)
    commands.push(`DROP TABLE IF EXISTS ${table.name}`);

  return commands;
}

function create(tables: Entity.Type[]){
  const commands = [];

  for(const table of tables){
    const { table: name } = table;
    const statements = [] as string[];

    table.fields.forEach(field => {
      const sql = column(field);

      if(sql)
        statements.push(sql);
    });

    commands.push(
      `CREATE TABLE IF NOT EXISTS ${name} (${statements.join(",")})`
    )
  }

  return commands;
}

function constraints(tables: Entity.Type[]){
  const commands = [] as string[];

  for(const table of tables){
    const statement = [] as string[];

    table.fields.forEach(field => {
      const { constraint } = field;

      if(constraint)
        statement.push(constraint);
    })

    if(statement.length)
      commands.push(`ALTER TABLE ${table.name} ${statement.join(", ")}`);
  }

  return commands;
}

function column(from: Field){
  if(from.datatype === undefined)
    return;

  let statement = `${qualify(from.column)} ${from.datatype}`;

  if(!from.nullable)
    statement += " NOT NULL";

  if(from.default !== undefined)
    statement += ` DEFAULT ${escapeString(from.default)}`;

  if(from.primary)
    statement += " PRIMARY KEY";

  if(from.datatype == "INTEGER" && from.increment)
    statement += " AUTOINCREMENT";

  return statement;
}