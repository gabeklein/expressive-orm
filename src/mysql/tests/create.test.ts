import { Bool, Entity, Int, Many, One, VarChar } from '../..';
import { bootstrap } from '../bootstrap';

it("will create tables", () => {
  class Author extends Entity {
    name = VarChar();
    age = Int();
    nickname = VarChar({ nullable: true });
    active = Bool();
    books = Many(Book);
    publisher = One(Publisher);
  }
  
  class Publisher extends Entity {
    name = VarChar();
  }
  
  class Book extends Entity {
    title = VarChar();
    author = One(Author);
    rating = Int();
  }

  const sql = bootstrap([
    Author,
    Book,
    Publisher
  ])

  expect(sql).toMatchSnapshot();
})