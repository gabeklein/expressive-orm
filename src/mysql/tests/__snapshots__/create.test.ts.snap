// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will create tables 1`] = `
	CREATE TABLE
	  IF NOT EXISTS Author (
	    \`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
	    \`name\` VARCHAR(255) NOT NULL,
	    \`age\` INT NOT NULL,
	    \`nickname\` VARCHAR(255),
	    \`active\` TINYINT NOT NULL,
	    \`publisherId\` INT NOT NULL
	  );
	
	CREATE TABLE
	  IF NOT EXISTS Book (
	    \`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
	    \`title\` VARCHAR(255) NOT NULL,
	    \`authorId\` INT NOT NULL,
	    \`rating\` INT NOT NULL
	  );
	
	CREATE TABLE
	  IF NOT EXISTS Publisher (
	    \`id\` INT NOT NULL AUTO_INCREMENT PRIMARY KEY,
	    \`name\` VARCHAR(255) NOT NULL
	  );
	
	ALTER TABLE
	  Author ADD CONSTRAINT \`FK_PublisherAuthor\` FOREIGN KEY (publisherId) REFERENCES Publisher (id);
	
	ALTER TABLE
	  Book ADD CONSTRAINT \`FK_AuthorBook\` FOREIGN KEY (authorId) REFERENCES Author (id)
`;
